elastic/kibana
# [Task Manager] Adding list of explicitly de-registered task types

Resolves https://github.com/elastic/kibana/issues/122389

## Summary

Adds a list of explicitly de-registered task types and updates claim script to only mark a task as `unrecognized` if the task type exists in that list.

In order to explicitly stop registering a task, plugin authors will need to remove the task registration code as well as add the task type to the `REMOVED_TYPES` list in `x-pack/plugins/task_manager/server/task_type_dictionary.ts`
* If the task is not registered but not added to `REMOVED_TYPES`, it will not be marked as `unrecognized`. It will just not be claimed.
* If a task is added to `REMOVED_TYPES` but still registered, an error will be thrown. These scenarios should be caught in development.

## Todo
* Add old siem rule type to list.

## To Verify

1. Create a rule that runs frequently and make sure it runs (I used the `People in Space` rule). Query for the rule in the task manager index and see that status is either `idle` or `running`.
2. Update the code so that this rule type is no longer registered with alerting

```
--- a/x-pack/examples/alerting_example/server/plugin.ts
+++ b/x-pack/examples/alerting_example/server/plugin.ts
@@ -28,7 +28,7 @@ export interface AlertingExampleDeps {
 export class AlertingExamplePlugin implements Plugin<void, void, AlertingExampleDeps> {
   public setup(core: CoreSetup, { alerting, features }: AlertingExampleDeps) {
     alerting.registerType(alwaysFiringAlert);
-    alerting.registerType(peopleInSpaceAlert);
+    // alerting.registerType(peopleInSpaceAlert);
```

3. Verify that the rule no longer runs and the status stays `idle`
4. Revert the update made in step 2. Verify that the rule starts running again.
5. Update the code to add this rule type to the list of removed types

```
--- a/x-pack/plugins/task_manager/server/task_type_dictionary.ts
+++ b/x-pack/plugins/task_manager/server/task_type_dictionary.ts
@@ -14,6 +14,7 @@ import { Logger } from '../../../../src/core/server';
 export const REMOVED_TYPES: string[] = [
   // for testing
   'sampleTaskRemovedType',
+  'alerting:example.people-in-space',
 ];
```

6. Verify that the rule status is updated to `unrecognized`.

### Checklist
